tensorflow/compiler/xla/service/gpu/matmul_utils.cc

```diff
@@ -752,53 +780,113 @@ Status MatmulPlan::DoMatmul(se::Stream* stream, se::DeviceMemoryBase a_buffer,
+      se::DeviceMemory<A>(a_buffer), se::DeviceMemory<B>(b_buffer),
+      se::HostOrDeviceScalar<Scale>(beta), se::DeviceMemory<C>(c_buffer),
       output, algorithm, scratch_allocator,
-      se::DeviceMemory<Input>(bias_buffer), profile_result);
+      se::DeviceMemory<Scale>(a_scale_buffer),
+      se::DeviceMemory<Scale>(b_scale_buffer),
+      se::DeviceMemory<Scale>(c_scale_buffer),
+      se::DeviceMemory<Scale>(d_scale_buffer),
+      se::DeviceMemory<Scale>(d_amax_buffer), se::DeviceMemory<C>(bias_buffer),
+      profile_result);
 }
 
 Status MatmulPlan::ExecuteOnStream(
     se::Stream* stream, se::DeviceMemoryBase a_buffer,
     se::DeviceMemoryBase b_buffer, se::DeviceMemoryBase c_buffer,
     se::DeviceMemoryBase d_buffer, se::DeviceMemoryBase bias_buffer,
+    se::DeviceMemoryBase a_scale_buffer, se::DeviceMemoryBase b_scale_buffer,
+    se::DeviceMemoryBase c_scale_buffer, se::DeviceMemoryBase d_scale_buffer,
+    se::DeviceMemoryBase d_amax_buffer,
     const se::cuda::BlasLt::MatmulAlgorithm& algorithm,
     se::ScratchAllocator& scratch_allocator,
     se::blas::ProfileResult* profile_result) const {
   if (must_swap_operands_) {
     std::swap(a_buffer, b_buffer);
   }
 
-  switch (plan_.d_desc.type()) {
-    case CUDA_R_16F:
-      return DoMatmul<Eigen::half, float>(stream, a_buffer, b_buffer, c_buffer,
-                                          d_buffer, bias_buffer, algorithm,
-                                          scratch_allocator, profile_result);
-    case CUDA_R_16BF:
-      return DoMatmul<Eigen::bfloat16, float>(
-          stream, a_buffer, b_buffer, c_buffer, d_buffer, bias_buffer,
-          algorithm, scratch_allocator, profile_result);
-    case CUDA_R_32F:
-      return DoMatmul<float>(stream, a_buffer, b_buffer, c_buffer, d_buffer,
-                             bias_buffer, algorithm, scratch_allocator,
-                             profile_result);
-    case CUDA_R_64F:
-      return DoMatmul<double>(stream, a_buffer, b_buffer, c_buffer, d_buffer,
-                              bias_buffer, algorithm, scratch_allocator,
-                              profile_result);
-    case CUDA_C_32F:
-      return DoMatmul<complex64>(stream, a_buffer, b_buffer, c_buffer, d_buffer,
-                                 bias_buffer, algorithm, scratch_allocator,
-                                 profile_result);
-    case CUDA_C_64F:
-      return DoMatmul<complex128>(stream, a_buffer, b_buffer, c_buffer,
-                                  d_buffer, bias_buffer, algorithm,
-                                  scratch_allocator, profile_result);
-    default:
-      return InternalError("Unexpected dtype");
+  std::tuple<cudaDataType_t, cudaDataType_t, cudaDataType_t, cudaDataType_t>
+      operand_types{plan_.a_desc.type(), plan_.b_desc.type(),
+                    plan_.c_desc.type(), plan_.d_desc.type()};
+
+  // FP8 allowed type combinations:
+  if (operand_types == std::make_tuple(CUDA_R_8F_E4M3, CUDA_R_8F_E4M3,
```

Comment:
Chatted offline: we were wondering how we can avoid code duplication. Here's a sketch of a workaround I could think of: https://godbolt.org/z/W7WMnohG6.

Reply:
Philipp tells me currently only 19 out of 256 template type combinations are legal, which need to be checked individually. The variant visitor suggestion I made was intended to eliminate the lengthy predicates and function calls, but if they have to be there anyway, his macro-based alternative that he has in mind might be a better compromise.